cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/preview/PreviewHttpHandler.java

```diff
@@ -146,4 +160,78 @@ public void getTracersData(HttpRequest request, HttpResponder responder,
+  }
+
+  @GET
+  @Path("/previews/{preview-id}/logs/prev")
+  public void getPreviewLogsPrev(HttpRequest request, HttpResponder responder,
+                                 @PathParam("namespace-id") String namespaceId,
+                                 @PathParam("preview-id") String previewId,
+                                 @QueryParam("max") @DefaultValue("50") int maxEvents,
+                                 @QueryParam("fromOffset") @DefaultValue("") String fromOffsetStr,
+                                 @QueryParam("escape") @DefaultValue("true") boolean escape,
+                                 @QueryParam("filter") @DefaultValue("") String filterStr,
+                                 @QueryParam("format") @DefaultValue("text") String format,
+                                 @QueryParam("suppress") List<String> suppress) throws Exception {
+    ProgramRunId runId = getProgramRunId(namespaceId, previewId);
+    RunRecordMeta runRecord = getRunRecord(namespaceId, previewId);
+    LoggingContext loggingContext =
+      LoggingContextHelper.getLoggingContextWithRunId(namespaceId, previewId, runId.getProgram(), runId.getType(),
+                                                      runId.getRun(), runRecord.getSystemArgs());
+    doPrev(responder, loggingContext, maxEvents, fromOffsetStr, escape, filterStr, runRecord, format, suppress);
+  }
+
+  @GET
+  @Path("/previews/{preview-id}/logs/next")
+  public void getPreviewLogsNext(HttpRequest request, HttpResponder responder,
+                                 @PathParam("namespace-id") String namespaceId,
+                                 @PathParam("preview-id") String previewId,
+                                 @QueryParam("max") @DefaultValue("50") int maxEvents,
+                                 @QueryParam("fromOffset") @DefaultValue("") String fromOffsetStr,
+                                 @QueryParam("escape") @DefaultValue("true") boolean escape,
+                                 @QueryParam("filter") @DefaultValue("") String filterStr,
+                                 @QueryParam("format") @DefaultValue("text") String format,
+                                 @QueryParam("suppress") List<String> suppress) throws Exception {
+    ProgramRunId runId = getProgramRunId(namespaceId, previewId);
+    RunRecordMeta runRecord = getRunRecord(namespaceId, previewId);
+    LoggingContext loggingContext =
+      LoggingContextHelper.getLoggingContextWithRunId(namespaceId, previewId, runId.getProgram(), runId.getType(),
+                                                      runId.getRun(), runRecord.getSystemArgs());
+    doNext(responder, loggingContext, maxEvents, fromOffsetStr, escape, filterStr, runRecord, format, suppress);
+  }
+
+
+  private ProgramRunId getProgramRunId(String namespaceId, String previewId) throws Exception {
+    ProgramRunId runId = previewManager.getRunner(new ApplicationId(namespaceId, previewId)).getProgramRunId();
+    if (runId == null) {
```

Comment:
No. Since when the preview status is DEPLOY_FAILED, the programRunId will not get initialized and be default it is null. Need this check here.